migrations/20231128_jellyfish_migration/utils/data_migration.go

```
@@ -0,0 +1,386 @@
+
+	capacity    *int
+	writeToDisk bool
+}
+
+func NewSettings(dryRun *bool, stopOnErr *bool, rollback *bool, rollbackSectionName *string, capacity *int, queryBatch *int, queryLimit *int, writeToDisk *bool) *Settings {
+	settings := &Settings{
+		writeToDisk:         false,
+		Rollback:            true,
+		RollbackSectionName: "_rollbackMigration",
+		DryRun:              true,
+		StopOnErr:           true,
+		QueryBatchSize:      50,
+		QueryBatchLimit:     100,
+	}
+	if dryRun != nil {
+		settings.DryRun = *dryRun
+	}
+	if stopOnErr != nil {
+		settings.StopOnErr = *stopOnErr
+	}
+	if rollback != nil {
+		settings.Rollback = *rollback
+	}
+	if rollbackSectionName != nil {
+		settings.RollbackSectionName = *rollbackSectionName
+	}
+	if writeToDisk != nil {
+		settings.writeToDisk = *writeToDisk
+	}
+	if queryBatch != nil {
+		settings.QueryBatchSize = *queryBatch
+	}
+	if queryLimit != nil {
+		settings.QueryBatchLimit = *queryLimit
+	}
+	if capacity != nil && *capacity > 0 {
+		settings.capacity = capacity
+		log.Printf("capped at %d items", *settings.capacity)
+	}
+	return settings
+}
+
+type DataMigration struct {
+	ctx                  context.Context
+	dataC                *mongo.Collection
+	settings             *Settings
+	updates              []mongo.WriteModel
+	groupedDiffs         map[string][]UpdateData
+	groupedErrors        groupedErrors
+	errorsCount          int
+	updatedCount         int
+	fetchedCount         int
+	lastUpdatedID        string
+	startedAt            time.Time
+	mongoInstanceChecker MongoInstanceCheck
+}
+
+type DataMigrationQueryFn func(m *DataMigration) bool
+type DataMigrationUpdateFn func(m *DataMigration) (int, error)
+
+type groupedErrors map[string][]ErrorData
+
+func NewMigration(ctx context.Context, settings *Settings, checker MongoInstanceCheck, dataC *mongo.Collection, lastID *string) (*DataMigration, error) {
+	var err error
+	if settings == nil {
+		err = errors.Join(err, errors.New("missing required settings"))
```

Why not just return the errors here rather than using a local variable?

---

BTW, do we want to check `dataC`, too?